docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,236 @@
+- kube-aggregator
+- kube-apiserver-basic-auth
+- kube-apiserver
+- service-account-key
+- ssh-keypair
+
+### Custom Resources and state of extension controllers
+
+Gardener deploys custom resources to the Host Seed cluster which are reconciled by extension controllers. The state of these controllers and any additional resources they create is independent of gardener and must also be migrated to the New Seed. Following is a list of custom resources, and the state which is generated by them that has to be migrated.
+
+- **BackupBucket**: nothing relevant for migration
+- **BackupEntry**: nothing relevant for migration
+- **ControlPlane**: nothing relevant for migration
+- **DNSProvider**/DNSEntry: nothing relevant for migration
+- **Extensions**: migration of state needs to be handled individually
+- **Infrastructure**: terraform state
+- **Network**: nothing relevant for migration
+- **OperatingSystemConfig**: nothing relevant for migration
+- **Worker**: Machine-Controller-Manager related objects: machineclasses, machinedeployments, machinesets, machines
+
+This list depends on the currently installed extensions and can change in the future
+
+## Proposal
+
+### Custom Resource on the garden cluster
+
+The Garden cluster has a new Custom Resource which is stored in the project namespace of the Shoot. It contains all the required data described above so that the control plane can be recreated on the __Destination Seed__.
+The metadata is separated into two sections. The first is generated by the Gardener Controller Manager and then either used to generate new resources (e.g secrets) or is directly deployed to the Shoot's control plane on the __Destination Seed__.
+
+The second is generated by the extension controllers in the seed and then synced to the shoot by a new ShootState Synchronization Controller (described below).
+
+```
+apiVersion: core.gardener.cloud/v1alpha1
+kind: ShootState
+metadata:
+  name: my-shoot
+  namespace: garden-core
+gardener:
+  secrets:
+  - name: ca
+    data:
+      ca.crt: ...
+      ca.key: ...
+  - name: ssh-keypair
+    data:
+      id_rsa: ...
+  - name:
+...
+extensions:
+- kind: Infrastructure
+  state: ... (Terraform state)
+- kind: ControlPlane
+  purpose: normal
+  state: ... (Certificates generated by the extension)
+- kind: Worker
+  state: ... (Machine objects)
+```
+
+The state data is saved as a `runtime.RawExtension` type, which can be encoded/decoded by the corresponding extension controller.
+
+There can be sensitive data in the `ShootState` which has to be hidden from the end-users. Currently we plan to keep the `ShootState` resource encrypted in etcd. This means that the Gardener API server must have an encryption configuration for `ShootState` resources and the Gardener Controller Manager must have access to the encryption key.
+
+An alternative is to have an additional namespace for every project called `garden-<project-name>-state` in which to keep the `ShootState` resource.
```

Comment:
I think we have agreed to not do this, so let's not mention it in the doc anymore.